frontend/apps/reader/modules/readerannotation.lua

```diff
@@ -0,0 +1,158 @@
+            if has_annotations then
+                config:saveSetting("annotations_paging", annotations)
+            end
+            annotations = config:readSetting("annotations_rolling") or {}
+            config:delSetting("annotations_rolling")
+        elseif self.ui.paging and annotations_type ~= "number" then
+            if has_annotations then
+                config:saveSetting("annotations_rolling", annotations)
+            end
+            annotations = config:readSetting("annotations_paging") or {}
+            config:delSetting("annotations_paging")
+        end
+        -- Build bookmarks/highlights
+        bookmarks, highlights = {}, {}
+        if #annotations > 0 then
+            local n = #annotations + 1
+            for i, an in ipairs(annotations) do
+                bookmarks[n - i] = self.buildBookmark(an)
+                if an.drawer then
+                    if highlights[an.pageno] == nil then
+                        highlights[an.pageno] = {}
+                    end
+                    table.insert(highlights[an.pageno], self.buildHighlight(an))
+                end
+            end
+        end
+    else -- old bookmarks/highlights
+        bookmarks = config:readSetting("bookmarks") or {}
+        highlights = config:readSetting("highlight") or {}
+        local has_bookmarks = #bookmarks > 0
+        local bookmarks_type = has_bookmarks and type(bookmarks[1].page)
+        if self.ui.rolling and bookmarks_type ~= "string" then
+            if has_bookmarks then
+                annotations = self:getAnnotationsFromBookmarksHighlights(bookmarks, highlights)
+                config:saveSetting("annotations_paging", annotations)
+            end
+            bookmarks = config:readSetting("bookmarks_rolling") or {}
+            highlights = config:readSetting("highlight_rolling") or {}
+            config:delSetting("bookmarks_rolling")
+            config:delSetting("highlight_rolling")
+        elseif self.ui.paging and bookmarks_type ~= "number" then
+            if has_bookmarks then
+                annotations = self:getAnnotationsFromBookmarksHighlights(bookmarks, highlights)
+                config:saveSetting("annotations_rolling", annotations)
+            end
+            bookmarks = config:readSetting("bookmarks_paging") or {}
+            highlights = config:readSetting("highlight_paging") or {}
+            config:delSetting("bookmarks_paging")
+            config:delSetting("highlight_paging")
+        end
+    end
+    self.ui.bookmark.bookmarks = bookmarks
+    self.view.highlight.saved = highlights
+end
+
+function ReaderAnnotation:onCloseDocument()
+    local annotations = self:getAnnotationsFromBookmarksHighlights(self.ui.bookmark.bookmarks, self.view.highlight.saved, true)
+    self.ui.doc_settings:saveSetting("annotations", annotations)
+    self.ui.doc_settings:delSetting("bookmarks")
+    self.ui.doc_settings:delSetting("highlight")
```

Comment:
May be these old `bookmarks/highlight_paging/rolling` could be saved under an other name (ie. `bookmarks_paging_obsolete`) instead of deleted, just for safety, and/or being able to manually resurrect them if needing to test/replay their migration.
(We don't really care if this crap stays in older book .sdr/metadata.lua.)